File: internal/topology/topology.go
package topology

import (
	"fmt"
	"os/exec"
	"strings"
	"testing"
)

type Topology struct {
	t      *testing.T
	Bridge string
	Nodes  []*Node
}

type Node struct {
	Index int
	NS    string // "ns_i2pN" или "ns_gnN"
	IP    string // "10.88.0.N+1"
	Veth  string // имя veth-интерфейса в namespace
}

func NewTopology(t *testing.T, n int, subnet string) *Topology {
	prefix := strings.ReplaceAll(subnet, ".", "_")
	bridgeName := fmt.Sprintf("br_%s", prefix)

	tp := &Topology{
		t:      t,
		Bridge: bridgeName,
		Nodes:  make([]*Node, n),
	}

	// cleanup registration
	t.Cleanup(func() {
		for i := n - 1; i >= 0; i-- {
			nsName := fmt.Sprintf("ns_%s%d", prefix, i)
			exec.Command("ip", "netns", "del", nsName).Run()
		}
		exec.Command("ip", "link", "set", bridgeName, "down").Run()
		exec.Command("ip", "link", "del", bridgeName, "type", "bridge").Run()
	})

	// 1. Create bridge
	tp.run("ip", "link", "add", bridgeName, "type", "bridge")
	tp.run("ip", "link", "set", bridgeName, "up")

	for i := 0; i < n; i++ {
		nsName := fmt.Sprintf("ns_%s%d", prefix, i)
		vethHost := fmt.Sprintf("veth_%s%d_h", prefix, i)
		vethNS := fmt.Sprintf("veth_%s%d_n", prefix, i)
		ipAddr := fmt.Sprintf("%s.%d/24", subnet, i+1)

		// 2. Create netns
		tp.run("ip", "netns", "add", nsName)
		tp.run("ip", "netns", "exec", nsName, "ip", "link", "set", "lo", "up")

		// 3. Create veth pair
		tp.run("ip", "link", "add", vethHost, "type", "veth", "peer", "name", vethNS)
		tp.run("ip", "link", "set", vethHost, "master", bridgeName)
		tp.run("ip", "link", "set", vethHost, "up")
		tp.run("ip", "link", "set", vethNS, "netns", nsName)

		// 4. Setup IP in NS
		tp.run("ip", "netns", "exec", nsName, "ip", "addr", "add", ipAddr, "dev", vethNS)
		tp.run("ip", "netns", "exec", nsName, "ip", "link", "set", vethNS, "up")

		tp.Nodes[i] = &Node{
			Index: i,
			NS:    nsName,
			IP:    fmt.Sprintf("%s.%d", subnet, i+1),
			Veth:  vethNS,
		}
	}

	return tp
}

func (tp *Topology) NodeIP(i int) string {
	if i < 0 || i >= len(tp.Nodes) {
		return ""
	}
	return tp.Nodes[i].IP
}

func (tp *Topology) run(name string, arg ...string) {
	cmd := exec.Command(name, arg...)
	if out, err := cmd.CombinedOutput(); err != nil {
		tp.t.Fatalf("command %s %v failed: %v\nOutput: %s", name, arg, err, string(out))
	}
}

File: testbed/src/topology.rs
use std::process::Command;

#[derive(Debug, Clone)]
pub struct NodeInfo {
    pub index: usize,
    pub ns: String,
    pub ip: String,
    pub veth: String,
}

pub struct Topology {
    bridge: String,
    pub nodes: Vec<NodeInfo>,
}

impl Topology {
    pub fn new(n: usize, subnet: &str) -> Self {
        let prefix = subnet.replace('.', "_");
        let bridge = format!("br_{prefix}");
        let mut tp = Self {
            bridge,
            nodes: Vec::with_capacity(n),
        };

        tp.run("ip", &["link", "add", &tp.bridge, "type", "bridge"]);
        tp.run("ip", &["link", "set", &tp.bridge, "up"]);

        for i in 0..n {
            let ns = format!("ns_{prefix}{i}");
            let veth_host = format!("veth_{prefix}{i}_h");
            let veth_ns = format!("veth_{prefix}{i}_n");
            let ip_cidr = format!("{subnet}.{}/24", i + 1);

            tp.run("ip", &["netns", "add", &ns]);
            tp.run(
                "ip",
                &["netns", "exec", &ns, "ip", "link", "set", "lo", "up"],
            );
            tp.run(
                "ip",
                &[
                    "link", "add", &veth_host, "type", "veth", "peer", "name", &veth_ns,
                ],
            );
            tp.run("ip", &["link", "set", &veth_host, "master", &tp.bridge]);
            tp.run("ip", &["link", "set", &veth_host, "up"]);
            tp.run("ip", &["link", "set", &veth_ns, "netns", &ns]);
            tp.run(
                "ip",
                &[
                    "netns", "exec", &ns, "ip", "addr", "add", &ip_cidr, "dev", &veth_ns,
                ],
            );
            tp.run(
                "ip",
                &["netns", "exec", &ns, "ip", "link", "set", &veth_ns, "up"],
            );

            tp.nodes.push(NodeInfo {
                index: i,
                ns,
                ip: format!("{subnet}.{}", i + 1),
                veth: veth_ns,
            });
        }

        tp
    }

    fn run(&self, name: &str, args: &[&str]) {
        let out = Command::new(name)
            .args(args)
            .output()
            .unwrap_or_else(|e| panic!("command {name} {args:?} failed to start: {e}"));
        if !out.status.success() {
            panic!(
                "command {name} {args:?} failed: {}\nstdout:\n{}\nstderr:\n{}",
                out.status,
                String::from_utf8_lossy(&out.stdout),
                String::from_utf8_lossy(&out.stderr)
            );
        }
    }
}

impl Drop for Topology {
    fn drop(&mut self) {
        for node in self.nodes.iter().rev() {
            let _ = Command::new("ip").args(["netns", "del", &node.ns]).status();
        }
        let _ = Command::new("ip")
            .args(["link", "set", &self.bridge, "down"])
            .status();
        let _ = Command::new("ip")
            .args(["link", "del", &self.bridge, "type", "bridge"])
            .status();
    }
}
